evm/vm/forks/sharding/vmc_handler.py

```diff
@@ -118,44 +103,79 @@ def _get_new_logs(self, shard_id):
     #       status of `new_logs`, `unchecked_logs`, and `current_score`, when it start to run
     #       `GUESS_HEAD`. If there is a new block arriving, just restore them to the saved status,
     #       append new logs to `new_logs`, and re-run `GUESS_HEAD`
-    def fetch_candidate_head(self, shard_id):
+    def fetch_candidate_head(self):
         # Try to return a log that has the score that we are checking for,
         # checking in order of oldest to most recent.
         unchecked_logs = pipe(
-            self.unchecked_logs[shard_id],
+            self.unchecked_logs,
             enumerate,
             tuple,
             reversed,
             tuple,
         )
-        current_score = self.current_score[shard_id]
+        current_score = self.current_score
 
         for idx, logs_entry in unchecked_logs:
             if logs_entry['score'] == current_score:
-                return self.unchecked_logs[shard_id].pop(idx)
+                return self.unchecked_logs.pop(idx)
         # If no further recorded but unchecked logs exist, go to the next
         # is_new_head = true log
         while True:
             # TODO: currently just raise when there is no log anymore
-            self.unchecked_logs[shard_id].append(self.get_next_log(shard_id))
-            if self.unchecked_logs[shard_id][-1]['is_new_head'] is True:
+            self.unchecked_logs.append(self.get_next_log())
+            if self.unchecked_logs[-1]['is_new_head'] is True:
                 break
-        log = self.unchecked_logs[shard_id].pop()
-        self.current_score[shard_id] = log['score']
+        log = self.unchecked_logs.pop()
+        self.current_score = log['score']
         return log
 
+
+class VMC(Contract):
+
+    logger = logging.getLogger("evm.chain.sharding.VMC")
+
+    def __init__(self, *args, default_privkey, **kwargs):
+        self.default_privkey = default_privkey
+        self.default_sender_address = default_privkey.public_key.to_canonical_address()
+        self.config = get_sharding_config()
+        self.shard_trackers = {}
+
+        super().__init__(*args, **kwargs)
+
+    def set_shard_tracker(self, shard_id, shard_tracker):
+        self.shard_trackers[shard_id] = shard_tracker
+
+    def get_shard_tracker(self, shard_id):
+        if shard_id not in self.shard_trackers:
+            raise ShardNotTracked('Shard {} is not tracked'.format(shard_id))
+        return self.shard_trackers[shard_id]
+
+    # TODO: currently just calls `shard_tracker.get_next_log`
+    def get_next_log(self, shard_id):
+        if shard_id not in self.shard_trackers:
+            raise ShardNotTracked('Shard {} is not tracked'.format(shard_id))
+        shard_tracker = self.shard_trackers[shard_id]
+        return shard_tracker.get_next_log()
+
+    # TODO: currently just calls `shard_tracker.fetch_candidate_head`
+    def fetch_candidate_head(self, shard_id):
+        if shard_id not in self.shard_trackers:
```

Comment:
Same here about removing this logic in favor of using `self.get_shard_tracker(shard_id)`

Reply:
Will change it. Thank you:)